community/kernel/src/test/java/org/neo4j/kernel/impl/store/TimeZoneMappingTest.java

```diff
@@ -0,0 +1,179 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store;
+
+import org.apache.commons.compress.archivers.ArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.junit.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.zip.GZIPInputStream;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.isEmptyString;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+public class TimeZoneMappingTest
+{
+    @Test
+    public void weSupportAllJavaZoneIds()
+    {
+        ZoneId.getAvailableZoneIds().forEach( s ->
+        {
+            short num = TimeZoneMapping.map( s );
+            assertThat( "Our time zone table does not have a mapping for " + s, num, greaterThanOrEqualTo( (short) 0 ) );
+
+            String nameFromTable = TimeZoneMapping.map( num );
+            if ( !s.equals( nameFromTable ) )
+            {
+                // The test is running on an older Java version and `s` has been removed since, thus it points to a different zone now.
+                // That zone should point to itself, however.
+                assertThat( "Our time zone table has inconsistent mapping for " + nameFromTable, TimeZoneMapping.map( TimeZoneMapping.map( nameFromTable ) ),
+                        equalTo( nameFromTable ) );
+            }
+        } );
+    }
+
+    @Test
+    public void weSupportDeletedZoneIdEastSaskatchewan()
+    {
+        try
+        {
+            short eastSaskatchewan = TimeZoneMapping.map( "Canada/East-Saskatchewan" );
+            assertThat( "Our time zone table does not remap Canada/East-Saskatchewan to Canada/Saskatchewan", TimeZoneMapping.map( eastSaskatchewan ),
+                    equalTo( "Canada/Saskatchewan" ) );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            fail( "Our time zone table does not support Canada/East-Saskatchewan" );
+        }
+    }
+
+    @Test
+    public void updateTimeZoneMappings() throws IOException
+    {
+        String foo = "https://www.iana.org/time-zones/repository/tzdata-latest.tar.gz";
+        List<String> versionsToUpgrade = new ArrayList<>();
+
+        try ( TarArchiveInputStream tar = new TarArchiveInputStream( new GZIPInputStream( new URL( foo ).openStream() ) ) )
```

Comment:
I'd like to factor this code into a some helper method:
```
void readTZDBFile( String release, String file, Predicate<String> consumer ) throws IOException
{
    String uri = release == null
            ? "https://data.iana.org/time-zones/tzdata-latest.tar.gz"
            : "https://data.iana.org/time-zones/releases/tzdata" + release + ".tar.gz";
    try ( TarArchiveInputStream tar = new TarArchiveInputStream( new GZIPInputStream( new URL( uri ).openStream() ) ) )
    {
        for ( ArchiveEntry entry; (entry = tar.getNextEntry()) != null; )
        {
            if ( file.equalsIgnoreCase( entry.getName() ) )
            {
                try ( BufferedReader reader = new BufferedReader( new InputStreamReader( tar ) ) )
                {
                    for ( String line; (line = reader.readLine()) != null; )
                    {
                        if ( !consumer.test( line ) )
                        {
                            break;
                        }
                    }
                }
                break;
            }
        }
    }
}
```

Then our two usages of this pattern would become:
```
readTZDBFile( null, "NEWS", line ->
{
    if ( line.startsWith( "Release " ) )
    {
        String substring = line.substring( line.indexOf( ' ' ) + 1 );
        String release = substring.substring( 0, substring.indexOf( ' ' ) );
        if ( TimeZoneMapping.LATEST_SUPPORTED_IANA_VERSION.equals( release ) )
        {
            return false; // stop reading
        }
        versionsToUpgrade.add( release );
    }
    return true; // continue reading
} );
```
and:
```
readTZDBFile( release, "zone1970.tab", line ->
{
    if ( !line.startsWith( "#" ) )
    {
        ianaSupportedTzs.add( line.split( "\\t" )[2] );
    }
    return true; // read all lines
} );
```

Which is much easier on the eyes (even if it does need the comments for explaining how the return value is used for control flow).

Reply:
Agreed